sycl/test-e2e/ESIMD/unified_memory_api/Inputs/copyto_copyfrom.hpp

```diff
@@ -0,0 +1,426 @@
+//==-- copyto_copyfrom.hpp - DPC++ ESIMD on-device test -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-------------------------------------------------------------------===//
+
+#include "common.hpp"
+
+using namespace sycl;
+using namespace sycl::ext::intel::esimd;
+
+// Returns true iff verification is passed.
+template <typename T> bool verify(T OutVal, const T *Out, size_t Size, int N) {
+  int NumErrors = 0;
+  using Tuint = sycl::_V1::ext::intel::esimd::detail::uint_type_t<sizeof(T)>;
+  for (int i = 0; i < Size; i++) {
+    Tuint Expected = sycl::bit_cast<Tuint>((T)(i + 3));
+    Tuint Computed = sycl::bit_cast<Tuint>(Out[i]);
+    if (Computed != Expected) {
+      NumErrors++;
+      if (NumErrors < 32)
+        std::cout << "Out[" << i << "] = " << std::to_string(Computed) << " vs "
+                  << std::to_string(Expected) << std::endl;
+    }
+  }
+  return NumErrors == 0;
+}
+
+template <typename T, uint16_t N, bool UseProperties, typename PropertiesT>
+bool testUSM(queue Q, uint32_t Groups, uint32_t Threads,
+             PropertiesT Properties) {
+
+  uint16_t Size = Groups * Threads * N;
+
+  std::cout << "USM case: T=" << esimd_test::type_name<T>() << ",N=" << N
+            << ",UseProperties=" << UseProperties << std::endl;
+
+  sycl::range<1> GlobalRange{Groups};
+  sycl::range<1> LocalRange{Threads};
+  sycl::nd_range<1> Range{GlobalRange * LocalRange, LocalRange};
+  constexpr size_t Alignment = getAlignment<T, N, false>(Properties);
+  T *Out = sycl::aligned_alloc_shared<T>(Alignment, Size, Q);
+  T OutVal = esimd_test::getRandomValue<T>();
+  for (int i = 0; i < Size; i++)
+    Out[i] = OutVal;
+
+  try {
+    Q.submit([&](handler &cgh) {
+       cgh.parallel_for(Range, [=](sycl::nd_item<1> ndi) SYCL_ESIMD_KERNEL {
+         uint16_t GlobalID = ndi.get_global_id(0);
+         uint32_t ElemOff = GlobalID * N;
+         simd<uint32_t, N> PassThruInt(ElemOff, 1);
+         simd<T, N> Vals = PassThruInt;
+
+         if constexpr (UseProperties) {
+           Vals.copy_to(Out + ElemOff, PropertiesT{});
+           Vals.copy_from(Out + ElemOff, PropertiesT{});
```

Comment:
done for USM and normal ACC, i didn't have a good idea how to do it with slm acc, if you have an idea let me know, for now I left it as is.

Reply:
Thank you.
It could be `scatter` to initialize 1st half of SLM and `copy_from` to load from it.
Similarly, `copy_to()` to initialize the 2nd part of SLM, then `gather` to load SLM. It would be cross-check eliminating the opting-out of store+load.

Reply:
thanks, i've implemented something i hope is similar to this in my latest commit, please take a look when you get a chance